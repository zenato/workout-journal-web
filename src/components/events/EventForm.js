import React from 'react';
import PropTypes from 'prop-types';
import { Field } from 'redux-form';
import styled from 'styled-components';
import { FormGroup, Button, Input } from 'components/form';

const Form = styled.form`
  padding: 10px;
`;

const Footer = styled.div`
  text-align: center;
  * {
    margin-left: 5px;
  }
`;

const Error = styled.div`
    color: #f00;
    text-align: center;
    font-size: 0.7rem;
    padding: 5px;
`;

const EventForm = ({ hasError, handleSubmit, initialValues, onDelete, onMoveList }) => (
  <Form onSubmit={handleSubmit}>
    <FormGroup label="Name">
      <Field type="text" name="name" component={Input} />
    </FormGroup>
    <FormGroup label="Unit">
      <Field type="text" name="unit" component={Input} />
    </FormGroup>
    <FormGroup label="Value">
      <Field type="text" name="value" component={Input} />
    </FormGroup>
    <FormGroup label="Remark">
      <Field type="text" name="remark" component={Input} />
    </FormGroup>

    {hasError && (
      <Error>
        <span>Oops, An expected error seems to have occurred.</span>
      </Error>
    )}

    <Footer>
      <Button type="submit" value="Save" className="primary" />
      <Button value="List" onClick={onMoveList} />
      {initialValues && <Button value="Delete" onClick={onDelete} />}
    </Footer>
  </Form>
);

EventForm.propTypes = {
  initialValues: PropTypes.object,
  hasError: PropTypes.bool.isRequired,
  handleSubmit: PropTypes.func.isRequired,
  onDelete: PropTypes.func.isRequired,
  onMoveList: PropTypes.func.isRequired,
};

export const validate = ({ name, unit, value }) => {
  const errors = {};

  if (!name) {
    errors.name = 'Required';
  }
  if (!unit) {
    errors.unit = 'Required';
  }
  if (!value) {
    errors.value = 'Required';
  } else if (!/^[0-9]{1,3}$/.test(value)) {
    errors.value = 'Must be a number, 0-999';
  }

  return errors;
};

export default EventForm;
