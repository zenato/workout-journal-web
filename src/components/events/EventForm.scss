@import 'common';

.event-form {
  padding: 10px;

  .item {
    display: flex;
    margin-bottom: 5px;
    align-items: center;

    label {
      flex: 2;
      text-align: right;
      font-weight: bold;
      margin-right: 5px;
    }

    .field {
      flex: 10;
      padding-right: 10px;
      vertical-align: middle;
    }

    .error-message {
      color: #f00;
      font-size: 0.7rem;
      padding: 5px;
    }
  }

  .tool {
    text-align: center;

    * {
      margin-left: 5px;
    }
  }

  .error {
    color: #f00;
    text-align: center;
    font-size: 0.7rem;
    padding: 5px;
  }
}

