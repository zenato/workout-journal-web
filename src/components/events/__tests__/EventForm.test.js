import React from 'react';
import { shallow } from 'enzyme';
import EventForm from '../EventForm';

describe('EventForm', () => {
  const props = {
    handleSubmit: jest.fn(),
    initialValues: null,
    error: null,
    onDelete: () => {},
    onMoveList: () => {},
  };

  it('should render error message if not empty `error` props.', () => {
    const component = shallow(<EventForm {...props} error={['error']} />);
    expect(component).toMatchSnapshot();
  });

  it('should render delete button if not empty `initialValues` props.', () => {
    const component = shallow(<EventForm {...props} initialValues={{ id: '1' }} />);
    expect(component).toMatchSnapshot();
  });

  it('call `handleSubmit` when submitting form.', () => {
    const component = shallow(<EventForm {...props} />);
    component.find('form').simulate('submit');
    expect(props.handleSubmit).toBeCalled();
  });
});
