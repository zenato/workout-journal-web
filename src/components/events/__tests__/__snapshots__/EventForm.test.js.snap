// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EventForm should render delete button if not empty \`initialValues\` props. 1`] = `
<form
  className="event-form"
  onSubmit={[Function]}
>
  <Field
    component={[Function]}
    label="Name"
    name="name"
    type="text"
  />
  <Field
    component={[Function]}
    label="Unit"
    name="unit"
    type="text"
  />
  <Field
    component={[Function]}
    label="Value"
    name="value"
    type="text"
  />
  <Field
    component={[Function]}
    label="Remark"
    name="remark"
    type="text"
  />
  <div
    className="tool"
  >
    <Button
      className="primary"
      type="submit"
      value="Save"
    />
    <Button
      onClick={[Function]}
      value="List"
    />
    <Button
      onClick={[Function]}
      value="Delete"
    />
  </div>
</form>
`;

exports[`EventForm should render error message if not empty \`error\` props. 1`] = `
<form
  className="event-form"
  onSubmit={[Function]}
>
  <Field
    component={[Function]}
    label="Name"
    name="name"
    type="text"
  />
  <Field
    component={[Function]}
    label="Unit"
    name="unit"
    type="text"
  />
  <Field
    component={[Function]}
    label="Value"
    name="value"
    type="text"
  />
  <Field
    component={[Function]}
    label="Remark"
    name="remark"
    type="text"
  />
  <div
    className="error"
  >
    <span>
      Oops, An expected error seems to have occurred.
    </span>
  </div>
  <div
    className="tool"
  >
    <Button
      className="primary"
      type="submit"
      value="Save"
    />
    <Button
      onClick={[Function]}
      value="List"
    />
  </div>
</form>
`;
